frame/election-providers/src/two_phase/mod.rs

```diff
@@ -0,0 +1,1622 @@
+	/// This checks the solution for the following:
+	///
+	/// 0. **all** of the used indices must be correct.
+	/// 1. present correct number of winners.
+	/// 2. any assignment is checked to match with [Snapshot::voters].
+	/// 3. for each assignment, the check of `ElectionDataProvider` is also examined.
+	/// 4. the claimed score is valid.
+	fn feasibility_check(
+		solution: RawSolution<CompactOf<T>>,
+		compute: ElectionCompute,
+	) -> Result<ReadySolution<T::AccountId>, FeasibilityError> {
+		let RawSolution {
+			compact,
+			score,
+			round,
+		} = solution;
+
+		// first, check round.
+		ensure!(Self::round() == round, FeasibilityError::InvalidRound);
+
+		// winners are not directly encoded in the solution.
+		let winners = compact.unique_targets();
+
+		// TODO: this can probably checked ahead of time, both in signed and unsigned story.
+		let desired_targets =
+			Self::desired_targets().ok_or(FeasibilityError::SnapshotUnavailable)?;
+
+		ensure!(
+			winners.len() as u32 == desired_targets,
+			FeasibilityError::WrongWinnerCount,
+		);
+
+		// read the entire snapshot.
+		let RoundSnapshot {
+			voters: snapshot_voters,
+			targets: snapshot_targets,
+		} = Self::snapshot().ok_or(FeasibilityError::SnapshotUnavailable)?;
+
+		// ----- Start building. First, we need some closures.
+		let voter_at = crate::voter_at_fn!(snapshot_voters, T::AccountId, T);
+		let target_at = crate::target_at_fn!(snapshot_targets, T::AccountId, T);
+
+		// first, make sure that all the winners are sane.
+		let winners = winners
+			.into_iter()
+			.map(|i| target_at(i).ok_or(FeasibilityError::InvalidWinner))
+			.collect::<Result<Vec<T::AccountId>, FeasibilityError>>()?;
+
+		// Then convert compact -> Assignment. This will fail if any of the indices are gibberish.
+		// that winner indices are already checked.
+		let assignments = compact
+			.into_assignment(voter_at, target_at)
+			.map_err::<FeasibilityError, _>(Into::into)?;
+
+		// Ensure that assignments is correct.
+		let _ = assignments
+			.iter()
+			.map(|ref assignment| {
+				// check that assignment.who is actually a voter (defensive-only).
+				// NOTE: This check is fully defensive as we know that this assignment.who is
+				// a voter, what we don't know is stake and targets. Essentially, we have lost the
+				// index after converting `compact` -> `assignment`. This can be optimized in the
+				// future.
+				let (_voter, _stake, targets) = snapshot_voters
+					.iter()
+					.find(|(v, _, _)| v == &assignment.who)
+					.ok_or(FeasibilityError::InvalidVoter)?;
+				// check that all of the targets are valid based on the snapshot.
+				if assignment
+					.distribution
+					.iter()
+					.any(|(d, _)| !targets.contains(d))
+				{
+					return Err(FeasibilityError::InvalidVote);
+				}
+				Ok(())
+			})
+			.collect::<Result<(), FeasibilityError>>()?;
+
+		// ----- Start building support. First, we need some more closures.
+		let stake_of = stake_of_fn!(snapshot_voters, T::AccountId);
```

Comment:
same for this function, basically we iterate on the whole voters vec to find the stake of the given voter.

Maybe we can just build a BTreeMap or at least sorted voters to really improve these iterations.

Reply:
This is something that I would actually consider doing in this PR. BTreeMap should be much better.